classes/Api/ThankYous.php

```diff
@@ -0,0 +1,809 @@
+		}
+
+		return $thank_yous_likes;
+	}
+
+	/**
+	 * @param SecurityContext $context
+	 * @param int[]|null      $user_ids
+	 * @param array|null      $date_range
+	 * @param int[]|null      $tag_ids
+	 * @return int
+	 */
+	public function GetTotalUsers(SecurityContext $context, ?array $user_ids = null, ?array $date_range = null, ?array $tag_ids = null): int
+	{
+		$extranet_ids = $this->GetVisibleExtranetIds($context);
+
+		return $this->thank_yous_repository->GetTotalUsers($user_ids, $date_range, $tag_ids, $extranet_ids);
+	}
+
+	public function GetTotalTags(SecurityContext $context, ?array $date_range = null, ?array $thanked_user_ids = null, ?array $tag_ids = null): int
+	{
+		$extranet_ids = $this->GetVisibleExtranetIds($context);
+
+		return $this->thank_yous_repository->GetTotalTags($extranet_ids, true, $date_range, $thanked_user_ids, $tag_ids);
+	}
+
+	/**
+	 * @param int|int[] $object_types_id
+	 * @return string|string[]
+	 * @throws ThankYouOClass - If the Name of the oClass could not be determined.
+	 */
+	public function GetOwnerClassNamesFromIds(array $object_types_id)
+	{
+		return $this->utility->GetOwnerClassNamesFromIds($object_types_id);
+	}
+
+	/**
+	 * Returns an array of Users indexed by their IDs.
+	 *
+	 * @param int[] $user_ids
+	 * @return User[]
+	 */
+	public function GetUsers(array $user_ids)
+	{
+		return $this->thank_yous_repository->GetUsers($user_ids);
+	}
+
+	/**
+	 * Given an array of arrays with offset id = int, oclass = int, returns a distinct list of User IDs.
+	 *
+	 * @param array $owner_classes
+	 * @return int[]
+	 */
+	public function GetDistinctUserIdsFromOwnerClasses(array $owner_classes)
+	{
+		$acl = $this->acl_repository->Get(0, 0);
+
+		foreach ($owner_classes as $owner_class_member)
+		{
+			$id        = $owner_class_member['id'] ?? null;
+			$oclass_id = $owner_class_member['oclass'] ?? null;
+
+			if (!isset($id))
+			{
+				throw new InvalidArgumentException("Failed to Get Distinct User IDs From Owner Classes, one or more Owner Class Members does not have an ID");
+			}
+			if (!is_int($id) || !($id > 0))
+			{
+				throw new InvalidArgumentException("Failed to Get Distinct User IDs From Owner Classes, Owner Class Member ID " . (string) $id . " is not a natural number");
+			}
+
+			if (!isset($oclass_id))
+			{
+				throw new InvalidArgumentException("Failed to Get Distinct User IDs From Owner Classes, one or more Owner Class Members does not have an Owner Class ID");
+			}
+			if (!is_int($oclass_id))
+			{
+				throw new InvalidArgumentException("Failed to Get Distinct User IDs From Owner Classes, Owner Class ID " . (string) $oclass_id . " is not an integer");
+			}
+
+			try
+			{
+				$acl->Add(0, $oclass_id, $id);
+			} catch (InvalidSubjectException $invalid_subject_exception)
+			{
+				throw new LogicException("Unexpected Exception thrown by Acl method Add", null, $invalid_subject_exception);
```

Comment:
`InvalidSubjectException` is already a `LogicException` (by way of extending `InvalidArgumentException`) so this does not need rethrowing at all.

Reply:
Fixed in 48d7138f0194ea7b048c6425539a637355592cbd